### PR TITLE
Registry: add a reusable function registry

### PR DESCRIPTION
**Pull request by ryantxu**
Part of #16756

This creates a generic `Extension` and `ExtensionRegistry` infrastructure that safely manages lazy initialising and access.

In this PR, it is just used for `fieldReducers`, but in #16756 will be used for transformations and matchers.

### CODE REVIEW

**Comment by ryantxu** [comment]
@torkelo -- now that we are on 6.4, I think this is ready to look at.

The one weird change that I am not sure about is moving `SelectOptionItem` from ui -> data package.  Since it is the data that the select option shows, I think that is OK, but admit that it is a little weird.

It looks like this PR also solves an issue with calculating allIsZero/allIsNull.  That should have been in its own issue and I did not realize I had it in here :(   I can pull it out, or keep it here

File: packages/grafana-data/src/types/select.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Used in select elements
+ */
+export interface SelectOptionItem<T = any> {
```
Comment (torkelo) [inline packages/grafana-data/src/types/select.ts]: it does feel strange to have it in data, but think it's ok. Maybe makes more sense if we change the name, the name feels a bit too UI and the doubling of Option & Item also feels a bit strange. 

SelectableItem ? 

Also remember that typescript has duck typing, you do not need to use the same type for things, just as long as it as a matching signature/props. So another alternative to moving this Type to data is to have two similar interfaces  (one defined here), and one defined in grafana/ui.

File: packages/grafana-data/src/utils/extensions.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+  /**
+   * Some extensions should not be user selectable
+   *  like: 'all' and 'any' matchers;
+   */
+  excludeFromPicker?: boolean;
+}
+
+interface ExtensionSelectInfo {
+  options: Array<SelectOptionItem<string>>;
```
Comment (torkelo) [inline packages/grafana-data/src/utils/extensions.ts]: Is it really needed to use the SelectOptionItem interface here? can't we define an ExtensionItem interface with things we need / use here? and then translate that to SelectOptionItem when we present the UI

**Review by torkelo — COMMENTED** [review]
Looks good, 2 thoughts / ideas for the SelectOptionItem problem

File: packages/grafana-data/src/types/select.ts (hunk 1 above)
Comment (ryantxu) [reply packages/grafana-data/src/types/select.ts]: I like SelectableItem 👍 

one interface seems better than two -- but yes, that is another reasonable option

File: packages/grafana-data/src/utils/extensions.ts (hunk 2 above)
Comment (ryantxu) [reply packages/grafana-data/src/utils/extensions.ts]: maybe... but we will want all the same fields, and the processing loop already happens here:
https://github.com/grafana/grafana/pull/17047/files#diff-e43a9d0466c5c7b68f77aefc4dab0762R81

but we could do it again.... or duck typing as you suggest

File: packages/grafana-data/src/types/select.ts (hunk 1 above)
Comment (torkelo) [reply packages/grafana-data/src/types/select.ts]: ok, lets keep it in @grafana/data in that case but rename it to SelectableItem

**Comment by ryantxu** [comment]
@torkelo -- I updated the PR.  Now the majority of the diff is:
```
- value?: SelectOptionItem<T>;
+ value?: SelectableItem<T>;
```

😃

**Comment by ryantxu** [comment]
@torkelo -- as I look at this again, what do you think about `SelectableValue` rather than `SelectableItem`?

**Comment by ryantxu** [comment]
@torkelo -- I updated the name changing SelectableItem to SelectableValue

We may want to avoid introducing something called "Extensions" since that should likely be reserved for a more substantial feature!

Maybe just call this a Registry?

`ExtensionRegistry<T extends Extension>` => `Registry<T extends RegistryItem>`?

**Comment by torkelo** [comment]
Registry feels better, Registry also feels a bit too generic, OptionRegistry ? The lack of name spacing feels bad in this case :) ) but can’t come up with a great alternative

**Comment by ryantxu** [comment]
We already use 'options' all over the place... mostly for react and function config passing, so I think we should avoid it if possible.

The registry will eventually hold:
* reducers
* transformers
* matchers (or whatever they end up being called)
* maybe [valueFormats](https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/utils/valueFormats/valueFormats.ts)

These are all functions of some sort so it could be `FunctionRegistry<T extends ????>`, but ???? should not be Function.

but there is nothing about the structure that should limit it to just functions, so I am back to proposing `Registry<T extends RegistryItem>`

**Comment by torkelo** [comment]
ok, lets go with Registry for now

**Review by torkelo — APPROVED** [review]

**Comment by gotjosh** [comment]
A head's up that this PR brought up is making master red. Putting together a quick PR on what I think it's the problem.

**Pull request merged**